common/src/main/java/com/hortonworks/registries/common/util/HdfsFileStorage.java

```diff
@@ -33,68 +42,146 @@
  */
 public class HdfsFileStorage implements FileStorage {
 
+    private static final Logger LOG = LoggerFactory.getLogger(HdfsFileStorage.class);
+
     // the configuration keys
     public static final String CONFIG_FSURL = "fsUrl";
     public static final String CONFIG_DIRECTORY = "directory";
+    public static final String CONFIG_KERBEROS_PRINCIPAL = "hdfs.kerberos.principal";
+    public static final String CONFIG_KERBEROS_KEYTAB = "hdfs.kerberos.keytab";
+    public static final Set<String> OWN_CONFIGS = ImmutableSet.of(CONFIG_FSURL, CONFIG_DIRECTORY, CONFIG_KERBEROS_PRINCIPAL, CONFIG_KERBEROS_KEYTAB);
 
-    private String fsUrl;
-    private String directory = DEFAULT_DIR;
-    private FileSystem hdfsFileSystem;
+    private String directory;
+    private Configuration hdfsConfig;
+    private URI fsUri;
+    private boolean kerberosEnabled = false;
 
     @Override
-    public void init(Map<String, String> config) {
-        Configuration hdfsConfig = new Configuration();
-        for(Map.Entry<String, String> entry: config.entrySet()) {
-            if(entry.getKey().equals(CONFIG_FSURL)) {
-                this.fsUrl = config.get(CONFIG_FSURL);
-            } else if(entry.getKey().equals(CONFIG_DIRECTORY)) {
-                this.directory = config.get(CONFIG_DIRECTORY);
-            } else {
-                hdfsConfig.set(entry.getKey(), entry.getValue());
-            }
+    public void init(Map<String, String> config) throws IOException {
+        String fsUrl = config.get(CONFIG_FSURL);
+        String kerberosPrincipal = config.get(CONFIG_KERBEROS_PRINCIPAL);
+        String keytabLocation = config.get(CONFIG_KERBEROS_KEYTAB);
+        directory = config.getOrDefault(CONFIG_DIRECTORY, DEFAULT_DIR);
+
+        hdfsConfig = new Configuration();
+
+        for(Map.Entry<String, String> entry:
+                Sets.filter(config.entrySet(), e -> !OWN_CONFIGS.contains(e.getKey()))) {
+            hdfsConfig.set(entry.getKey(), entry.getValue());
         }
 
         // make sure fsUrl is set
-        if(fsUrl == null) {
-            throw new RuntimeException("fsUrl must be specified for HdfsFileStorage.");
+        Preconditions.checkArgument(fsUrl != null, "fsUrl must be specified for HdfsFileStorage.");
+
+        Preconditions.checkArgument(keytabLocation != null || kerberosPrincipal == null,
+            "%s is needed when %s (== %s) is specified.",
+            CONFIG_KERBEROS_KEYTAB, CONFIG_KERBEROS_PRINCIPAL, kerberosPrincipal);
+
+        Preconditions.checkArgument(kerberosPrincipal != null || keytabLocation == null,
+            "%s is needed when %s (== %s) is specified.",
+            CONFIG_KERBEROS_PRINCIPAL, CONFIG_KERBEROS_KEYTAB, keytabLocation);
+
+        if (kerberosPrincipal != null) {
+            LOG.info("Logging in as kerberos principal {}", kerberosPrincipal);
+            UserGroupInformation.loginUserFromKeytab(kerberosPrincipal, keytabLocation);
```

Comment:
Not sure What's the best practice? Should we also check if the `getLoginUser` is null before we login. @sraju told me that UGI could be used in multiple parts of SR.

Reply:
- This is a one time setup, so I don't think I should check.
- Also kerberos user / principal is in the FS related part of the config (will be added with the reladed CSD change)
- Do you mean we should login with different principals to do different things in Schema Registry?

Reply:
Yeah, UGI will be used in multiple places going forward. In SR HA mechanism and Ranger Authorization. We should probably move the UGI initialization logic to RegistryApplication and move the Kerberos principal and keyTab configuration out of FS config. What do you guys think? @benyoka and @guruchai

Reply:
Should it be handled in upcoming releases - given the timeframe we need to deliver this? Currently loading the keytab depends on hadoop's security settings, probably this will have to change later too.

Reply:
Cool, lets handle this later.